torch/csrc/jit/autodiff.cpp

```diff
@@ -59,4 +57,170 @@ void differentiate(std::shared_ptr<Graph>& graph) {
+  // and used in the second one. They will need to be added as inputs of the reverse
+  // graph, and some of them may also need to be appended as outputs of the primal graph.
+  std::unordered_set<Value*> extra_reverse_inputs;
+  for (Node * node : *graph) {
+    if (node->stage() == 0) continue;
+    for (Value * input : node->inputs()) {
+      if (input->stage() != 0) continue;
+      extra_reverse_inputs.insert(input);
+    }
+  }
+
+  // --------------------------------------------------------------------------
+  // 2. Establish an ordering on inputs/outputs.
+  // --------------------------------------------------------------------------
+  // We will need to sort primal outputs and reverse inputs, so that they maintain
+  // an order of:
+  //   [primal outputs], [primal temporaries], [reverse inputs]
+  // where each section is sorted in the same way as in the input graph.
+  //
+  // This corresponds exactly to taking a topological ordering on values,
+  // but artificially boosting outputs of the first stage to precede all
+  // other values.
+  auto topo_order = computeValueTopoOrder(graph);
+  auto io_compare = [&](Value *a, Value *b) { return topo_order.at(a) < topo_order.at(b); };
+  { // Boost outputs of primal graph
+    int64_t next_nr = -static_cast<int64_t>(graph->outputs().size());
+    for (Value * output : graph->outputs()) {
+      if (output->stage() != 0) break;
+      topo_order[output] = next_nr++;
+    }
+  }
+
+  // --------------------------------------------------------------------------
+  // 3. Prepare input/outputs lists for both graphs.
+  // --------------------------------------------------------------------------
+  // It's simple to construct primal_inputs/reverse_outputs,
+  // but primal_outputs/reverse_inputs are much more subtle.
+  // Here's a summary of how they are supposed to look like:
+  //
+  // Primal outputs:
+  //   [original outputs], [temporaries]
+  //
+  // Reverse inputs:
+  //   [original outputs#], [temporaries], [output vjps (aka grad_outputs)]
+  //
+  // # this is an arbitrary subset of original outputs, that maintains the original ordering.
+  //   IMPORTANT: THIS MEANS THAT YOU CAN'T JUST TAKE OUTPUTS OF PRIMAL GRAPH AND FEED THEM
+  //              AS THE INPUTS OF THE REVERSE GRAPH.
+
+  // XXX: Take care when handling primal_outputs here - you can't put them in a set and
+  // just take them from it, because some values could have appeared multiple times in the
+  // outputs list, and this will deduplicate them.
+
+  value_list primal_inputs   = filter(graph->inputs(),  is_stage_0);
+  value_list primal_outputs  = filter(graph->outputs(), is_stage_0);
+  value_list reverse_inputs  = filter(graph->inputs(),  is_stage_1);
+  value_list reverse_outputs = filter(graph->outputs(), is_stage_1);
+
+  // Cache some information that will be useful in stage 4.
+  auto num_original_primal_outputs = primal_outputs.size();
```

Comment:
Too many random numbers used to slice arrays here.

Reply:
Random numbers? I'm only using the number of inputs/outputs, and this code doesn't slice anything.